vault/request_forwarding.go

```diff
@@ -134,197 +81,156 @@ func (c *Core) startForwarding(ctx context.Context) error {
-
-				// Accept the connection
-				conn, err := tlsLn.Accept()
-				if err != nil {
-					if err, ok := err.(net.Error); ok && !err.Timeout() {
-						c.logger.Debug("non-timeout error accepting on cluster port", "error", err)
-					}
-					if conn != nil {
-						conn.Close()
-					}
-					continue
-				}
-				if conn == nil {
-					continue
-				}
-
-				// Type assert to TLS connection and handshake to populate the
-				// connection state
-				tlsConn := conn.(*tls.Conn)
-
-				// Set a deadline for the handshake. This will cause clients
-				// that don't successfully auth to be kicked out quickly.
-				// Cluster connections should be reliable so being marginally
-				// aggressive here is fine.
-				err = tlsConn.SetDeadline(time.Now().Add(30 * time.Second))
-				if err != nil {
-					if c.logger.IsDebug() {
-						c.logger.Debug("error setting deadline for cluster connection", "error", err)
-					}
-					tlsConn.Close()
-					continue
-				}
-
-				err = tlsConn.Handshake()
-				if err != nil {
-					if c.logger.IsDebug() {
-						c.logger.Debug("error handshaking cluster connection", "error", err)
-					}
-					tlsConn.Close()
-					continue
-				}
-
-				// Now, set it back to unlimited
-				err = tlsConn.SetDeadline(time.Time{})
-				if err != nil {
-					if c.logger.IsDebug() {
-						c.logger.Debug("error setting deadline for cluster connection", "error", err)
-					}
-					tlsConn.Close()
-					continue
-				}
-
-				switch tlsConn.ConnectionState().NegotiatedProtocol {
-				case requestForwardingALPN:
-					if !ha {
-						tlsConn.Close()
-						continue
-					}
-
-					c.logger.Debug("got request forwarding connection")
-
-					shutdownWg.Add(2)
-					// quitCh is used to close the connection and the second
-					// goroutine if the server closes before closeCh.
-					quitCh := make(chan struct{})
-					go func() {
-						select {
-						case <-quitCh:
-						case <-closeCh:
-						}
-						tlsConn.Close()
-						shutdownWg.Done()
-					}()
-
-					go func() {
-						fws.ServeConn(tlsConn, &http2.ServeConnOpts{
-							Handler: fwRPCServer,
-							BaseConfig: &http.Server{
-								ErrorLog: c.logger.StandardLogger(nil),
-							},
-						})
-						// close the quitCh which will close the connection and
-						// the other goroutine.
-						close(quitCh)
-						shutdownWg.Done()
-					}()
-
-				case PerformanceReplicationALPN, DRReplicationALPN, perfStandbyALPN:
-					handleReplicationConn(ctx, c, shutdownWg, closeCh, fws, perfStandbyReplicationRPCServer, perfStandbyCache, tlsConn)
-				default:
-					c.logger.Debug("unknown negotiated protocol on cluster port")
-					tlsConn.Close()
-					continue
-				}
-			}
-		}()
+	return nil, nil
+}
+
+// ServerLookup satisfies the ClusterHandler interface and returns the server's
+// tls certs.
+func (rf *requestForwardingHandler) ServerLookup(ctx context.Context, clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	currCert := rf.core.localClusterCert.Load().([]byte)
```

Comment:
`rf.core.localClusterCert.Load()` can return `nil`. Should we check that? Otherwise, it might panic.

Reply:
Because we set a typed nil value we won't panic here https://github.com/hashicorp/vault/blob/9a980d34c75dd0760cbf4e75152763915b6aa3e7/vault/cluster.go#L101

It's a nice way to not require a type check with atomic values. We do nil check it bellow before we start using the value

Reply:
Ah cool. Thanks for the clarification.